deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java

```diff
@@ -205,7 +229,14 @@ void configureRuntimeProperties(
       CassandraClientBuildTimeConfig buildTimeConfig,
       Capabilities capabilities) {
     recorder.configureRuntimeProperties(runtimeConfig);
+    configureMetrics(recorder, buildTimeConfig, capabilities);
+    recorder.configureCompression(buildTimeConfig.protocolCompression);
+  }
 
+  private void configureMetrics(
+      CassandraClientRecorder recorder,
+      CassandraClientBuildTimeConfig buildTimeConfig,
+      Capabilities capabilities) {
```

Comment:
Maybe it's just me but I find the relationship between the processor and the recorder to be somewhat unclear.  I'm wondering if we shouldn't more clearly define roles here, specifically:

* The recorder exposes methods which take raw bits of data (most likely config data) and constructs what it needs internally to subsequently necessary to create CqlSessions
* The processor evaluates conditions at runtime and extracts/computes/otherwise obtains those raw bits of data.  It then uses the methods supplied by the recorder to setup what is needed.

Specifically, the processor shouldn't be responsible for constructing things like MetricsConfig instances.  It's job is to tell the recorder to either enable metrics (for a specified set of node and/or session metrics) or disable metrics.  The recorder is responsible for actually implementing that logic.  I'm envisioning something more like the following:

```java
  @Record(RUNTIME_INIT)
  @BuildStep
  void configureRuntimeProperties(
      CassandraClientRecorder recorder,
      CassandraClientConfig runtimeConfig,
      CassandraClientBuildTimeConfig buildTimeConfig,
      Capabilities capabilities) {

    recorder.runtimeProperties(runtimeConfig).setCompression(buildTimeConfig.protocolCompression);

    if (buildTimeConfig.metricsEnabled) {
      recorder.enableMetrics(
          buildTimeConfig.metricsNodeEnabled, buildTimeConfig.metricsSessionEnabled);
    } else {
      recorder.disableMetrics();
    }

    if (buildTimeConfig.metricsEnabled && capabilities.isCapabilityPresent(Capabilities.METRICS)) {
      recorder.setInjectedMetricRegistry();
    } else {
      recorder.setNoopMetricRegistry();
    }
  }
```

I've added some method chaining to further simplify the code but you get the general idea I'm going for here.

My suspicion is that as we add support for additional features a clear delineation of roles here will help us keep track of what's going on in this code.  Note that this is separate from the question of whether protocol compression should even have it's own method in the recorder vs. just being added to CassandraClientConnectionConfig.

Full diff supporting the snippet I showed above follows:

```
diff --git a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
index 70be4e2..3ac96cb 100644
--- a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
+++ b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
@@ -65,7 +65,6 @@ import io.quarkus.smallrye.health.deployment.spi.HealthBuildItem;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.Optional;
 import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Default;
 import javax.enterprise.inject.Produces;
@@ -228,21 +227,14 @@ class CassandraClientProcessor {
       CassandraClientConfig runtimeConfig,
       CassandraClientBuildTimeConfig buildTimeConfig,
       Capabilities capabilities) {
-    recorder.configureRuntimeProperties(runtimeConfig);
-    configureMetrics(recorder, buildTimeConfig, capabilities);
-    recorder.configureCompression(buildTimeConfig.protocolCompression);
-  }
 
-  private void configureMetrics(
-      CassandraClientRecorder recorder,
-      CassandraClientBuildTimeConfig buildTimeConfig,
-      Capabilities capabilities) {
+    recorder.runtimeProperties(runtimeConfig).setCompression(buildTimeConfig.protocolCompression);
+
     if (buildTimeConfig.metricsEnabled) {
-      recorder.configureMetrics(
-          new MetricsConfig(
-              buildTimeConfig.metricsNodeEnabled, buildTimeConfig.metricsSessionEnabled, true));
+      recorder.enableMetrics(
+          buildTimeConfig.metricsNodeEnabled, buildTimeConfig.metricsSessionEnabled);
     } else {
-      recorder.configureMetrics(new MetricsConfig(Optional.empty(), Optional.empty(), false));
+      recorder.disableMetrics();
     }
 
     if (buildTimeConfig.metricsEnabled && capabilities.isCapabilityPresent(Capabilities.METRICS)) {
diff --git a/runtime/src/main/java/com/datastax/oss/quarkus/runtime/CassandraClientRecorder.java b/runtime/src/main/java/com/datastax/oss/quarkus/runtime/CassandraClientRecorder.java
index 1efbf1c..53f8d0e 100644
--- a/runtime/src/main/java/com/datastax/oss/quarkus/runtime/CassandraClientRecorder.java
+++ b/runtime/src/main/java/com/datastax/oss/quarkus/runtime/CassandraClientRecorder.java
@@ -24,6 +24,8 @@ import io.quarkus.arc.runtime.BeanContainerListener;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.runtime.annotations.Recorder;
 import io.smallrye.metrics.MetricRegistries;
+import java.util.List;
+import java.util.Optional;
 import javax.enterprise.inject.Default;
 import javax.enterprise.util.AnnotationLiteral;
 import org.eclipse.microprofile.metrics.MetricRegistry;
@@ -36,10 +38,11 @@ public class CassandraClientRecorder {
     return beanContainer -> beanContainer.instance(cqlSessionProducerClass);
   }
 
-  public void configureRuntimeProperties(CassandraClientConfig config) {
+  public CassandraClientRecorder runtimeProperties(CassandraClientConfig config) {
     AbstractCassandraClientProducer producer =
         Arc.container().instance(AbstractCassandraClientProducer.class).get();
     producer.setCassandraClientConfig(config);
+    return this;
   }
 
   private AnnotationLiteral<Default> defaultName() {
@@ -50,10 +53,23 @@ public class CassandraClientRecorder {
     return new RuntimeValue<>(Arc.container().instance(CqlSession.class, defaultName()).get());
   }
 
-  public void configureMetrics(MetricsConfig metricsConfig) {
-    AbstractCassandraClientProducer producer =
+  public CassandraClientRecorder enableMetrics(
+      Optional<List<String>> nodeMetrics, Optional<List<String>> sessionMetrics) {
+
+    return setMetricConfig(new MetricsConfig(nodeMetrics, sessionMetrics, true));
+  }
+
+  public CassandraClientRecorder disableMetrics() {
+
+    return setMetricConfig(new MetricsConfig(Optional.empty(), Optional.empty(), false));
+  }
+
+  private CassandraClientRecorder setMetricConfig(MetricsConfig config) {
+
+    AbstractCassandraClientProducer rv =
         Arc.container().instance(AbstractCassandraClientProducer.class).get();
-    producer.setMetricsConfig(metricsConfig);
+    rv.setMetricsConfig(config);
+    return this;
   }
 
   public void setInjectedMetricRegistry() {
@@ -69,7 +85,7 @@ public class CassandraClientRecorder {
     producer.setMetricRegistry(new NoopMetricRegistry());
   }
 
-  public void configureCompression(String protocolCompression) {
+  public void setCompression(String protocolCompression) {
     AbstractCassandraClientProducer producer =
         Arc.container().instance(AbstractCassandraClientProducer.class).get();
     producer.setProtocolCompression(protocolCompression);
```

Reply:
Actually, this is another way around. The processor logic must be in the build time.
I've taken then the pattern of Recorder, Processor and their responsibilities from other extensions. Regarding returns `this`, I don't see a clear benefit of doing this. The recorder is not a builder. It is a way to integrate the build time code with runtime.

Reply:
I'm not sure I understand what you mean @tomekl007.  Nothing in the build-time/run-time distinction forces you to build MetricsConfig instances in the processor rather than the recorder.  In fact you're already doing what I'm suggesting with the metric registry.  In that case you call a method on the recorder and that method is responsible for actually creating the instance.  All I'm suggesting is making this pattern uniform.

Returning this is primarily about simplifying the code.  It's something I'd argue for personally but it's not essential by any means.

Reply:
I think that we can continue the discussion about settings in this ticket:
https://datastax-oss.atlassian.net/browse/JAVA-2697
Please add your thought about shape of the configuration mechanism there.